Issue #567: define loss as classes and use registry

mathpluscode opened the pull request:
# Description

Fixes #567

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

mathpluscode commented:
@YipengHu @acasamitjana @zacbaum @s-sd Would you have time review this PR ^^

In general, I
- refactored loss functions into classes
- merged some losses (like dice, dice_general) to make them more general, less redundant
- simplified the config so that later it will be easier to support multiple losses
- modified/refactored some unit-tests to ensure the code coverage

YipengHu commented:
Please check if docs that need to update...

mathpluscode commented:
> Please check if docs that need to update...

I don't think so, as there were no docs and @NMontanaBrown was writing it ^^ in #559.
I will have a check anyway.

YipengHu commented:
This looks good! I will send my comments/reviews in the next few days. The main concern is that the APIs are technically unchanged, while the config files are no longer back-compatible. What is the best way to inform users - i would even consider a simple script to convert from the old format to new one? any idea? @zacbaum @NMontanaBrown

mathpluscode commented:
> This looks good! I will send my comments/reviews in the next few days. The main concern is that the APIs are technically unchanged, while the config files are no longer back-compatible. What is the best way to inform users - i would even consider a simple script to convert from the old format to new one? any idea? @zacbaum @NMontanaBrown

do we have users....?
As it's not released, I suggest to wait until @NMontanaBrown's branch got merged, so we have full docs on config. Then we work on a better error handling. In case of the wrong config, the err message should be clear and point user to a certain page.

mathpluscode commented:
@zacbaum if you have time for this huge PR :) but no rush ^^

zacbaum commented:
Thanks for doing this @mathpluscode - It was clearly lots of work! All looks good as is; though I agree with @YipengHu:

> The main concern is that the APIs are technically unchanged, while the config files are no longer back-compatible. 

Is there no (non-trivial) to support both? I know that this likely means lots of duplicate code... But maybe better than just ridding the codebase of the previous way that things were done. Otherwise, this is a borderline major change as far as semver goes.

> What is the best way to inform users - i would even consider a simple script to convert from the old format to new one? any idea? 

I'd agree with this; or we deprecate it but leave it in the codebase for the time being and log a warning to the user and provide a link/info about how to update their config file. Then remove it in a later version. Allow any possible users the chance to update before we just remove it entirely.

mathpluscode commented:
> Thanks for doing this @mathpluscode - It was clearly lots of work! All looks good as is; though I agree with @YipengHu:
> 
> > The main concern is that the APIs are technically unchanged, while the config files are no longer back-compatible.
> 
> Is there no (non-trivial) to support both? I know that this likely means lots of duplicate code... But maybe better than just ridding the codebase of the previous way that things were done. Otherwise, this is a borderline major change as far as semver goes.
> 
> > What is the best way to inform users - i would even consider a simple script to convert from the old format to new one? any idea?
> 
> I'd agree with this; or we deprecate it but leave it in the codebase for the time being and log a warning to the user and provide a link/info about how to update their config file. Then remove it in a later version. Allow any possible users the chance to update before we just remove it entirely.

Agreed!


> What is the best way to inform users - i would even consider a simple script to convert from the old format to new one? any idea?

Considering/Assuming not having hundreds of users, I would prefer not "depreciating" the configs... it's too much tech debt to take care of. My proposition is **the well-organized error message**, so that if the user is using a wrong/outdated config, then the error message will point user to the docs and name clearly what should be the config. The problem here is: in `main` we do not have docs for config and @NMontanaBrown is writing them.

In short, before the next release, which would be in Jan-Feb, we would need to revise the changes and make sure the docs are up-to-date and clear.


What do you think? @zacbaum @YipengHu As there are still some more to refactor, e.g. #557, which will also bring some changes of configs. I intend to provide jupyter notebook or scripts like demo to showcase how to use custom model/loss/backbone etc in the end, and plus a detailed config documentation, plus well written error message, I guess it would be easy for user to debug.

zacbaum commented:
> Considering/Assuming not having hundreds of users, I would prefer not "depreciating" the configs... it's too much tech debt to take care of. My proposition is **the well-organized error message**, so that if the user is using a wrong/outdated config, then the error message will point user to the docs and name clearly what should be the config. The problem here is: in `main` we do not have docs for config and @NMontanaBrown is writing them.
> 
> In short, before the next release, which would be in Jan-Feb, we would need to revise the changes and make sure the docs are up-to-date and clear.

I'd prefer having the docs added with (or before) this PR if possible.

> What do you think? @zacbaum @YipengHu As there are still some more to refactor, e.g. #557, which will also bring some changes of configs. I intend to provide jupyter notebook or scripts like demo to showcase how to use custom model/loss/backbone etc in the end, and plus a detailed config documentation, plus well written error message, I guess it would be easy for user to debug.

If more to come, then we need to add warning if things are not as they used to be, and be sure that docs are updated with the changes; not after the fact.

YipengHu commented:
> > Considering/Assuming not having hundreds of users, I would prefer not "depreciating" the configs... it's too much tech debt to take care of. My proposition is **the well-organized error message**, so that if the user is using a wrong/outdated config, then the error message will point user to the docs and name clearly what should be the config. The problem here is: in `main` we do not have docs for config and @NMontanaBrown is writing them.
> > In short, before the next release, which would be in Jan-Feb, we would need to revise the changes and make sure the docs are up-to-date and clear.
> 
> I'd prefer having the docs added with (or before) this PR if possible.
> 
> > What do you think? @zacbaum @YipengHu As there are still some more to refactor, e.g. #557, which will also bring some changes of configs. I intend to provide jupyter notebook or scripts like demo to showcase how to use custom model/loss/backbone etc in the end, and plus a detailed config documentation, plus well written error message, I guess it would be easy for user to debug.
> 
> If more to come, then we need to add warning if things are not as they used to be, and be sure that docs are updated with the changes; not after the fact.

agreed - need to sync with @NMontanaBrown see how much she progressed. Then have some decent doc/error-messaging with the PR.

mathpluscode commented:
Then would you check if the changes here in this PR is ok for you? @zacbaum @YipengHu 
For configs, I will sync with Nina.

We merge this branch so that I can continue forward with models :)

mathpluscode commented:
@zacbaum I have a problem with the download test. It is downloading the main branch, but I modified the configs in my branch.... so that the test will not pass lol. Means that the test should depend on the current branch....

YipengHu commented:
> Then would you check if the changes here in this PR is ok for you? @zacbaum @YipengHu
> For configs, I will sync with Nina.
> 
> We merge this branch so that I can continue forward with models :)

what models?

mathpluscode commented:
> > Then would you check if the changes here in this PR is ok for you? @zacbaum @YipengHu
> 
> > For configs, I will sync with Nina.
> 
> > 
> 
> > We merge this branch so that I can continue forward with models :)
> 
> 
> 
> what models?

All models, ddf dvf etc, see issue #557

YipengHu commented on deepreg/model/loss/image.py:
```diff
@@ -3,44 +3,119 @@
-    """
-    Returns the calculated dissimilarity for each batch.
+@REGISTRY.register_loss(name="ssd")
+class SumSquaredDistance(tf.keras.losses.Loss):
```
it is better to be SumSquaredDifferences or  SumSquaredDiff? - distance can be confusing

mathpluscode replied:
Ok

mathpluscode replied:
done

YipengHu commented on deepreg/model/loss/image.py:
```diff
@@ -50,167 +125,157 @@ def local_normalized_cross_correlation(
+
+
+@REGISTRY.register_loss(name="lncc")
+class LocalNormalizedCrossCorrelation3DLoss(
```
class names are not consistent: some with "3D", some without (energy/image/label all can be dimension-sensitive), while this is with "3DLoss"

mathpluscode replied:
Those ones without 3D because they do not assume tensors have to be 3D.

YipengHu replied:
gradient and bending do assume 3d...

mathpluscode replied:
oh there are some inconsistencies. I've fixed them :)

YipengHu commented on deepreg/model/loss/util.py:
```diff
@@ -0,0 +1,10 @@
+import tensorflow as tf
+
+
+class NegativeLossMixin(tf.keras.losses.Loss):
```
do we really need a mixin class? it hurts readability to me.

mathpluscode replied:
Yes we need. As user you do not need to care about mixin.

YipengHu commented on deepreg/registry.py:
```diff
@@ -74,11 +74,50 @@ def decorator(_cls):
 
         return decorator
 
+    def build_from_config(self, category: str, config: dict, default_args=None):
```
this is probably the place to add a few checks to produce back-compatible error messages? you can just check a few old ones, e.g. energy_type, loss_type, loss_scale... then i think it should be pretty safe to recognise the old format config files.

mathpluscode replied:
I strongly object the back compatibility.

mathpluscode replied:
The whole idea is to make it simple. I will do meaningful error messages.

YipengHu replied:
> The whole idea is to make it simple. I will do meaningful error messages.

i didn't mean back-compatible code....

YipengHu commented on demos/grouped_mask_prostate_longitudinal/grouped_mask_prostate_longitudinal.yaml:
```diff
@@ -21,21 +21,16 @@ train:
     regularization:
       weight: 0.01 # weight of regularization loss
-      energy_type: "gradient-l2" # value should be bending / gradient-l1 / gradient-l2
+      name: "gradient" # options include "bending", "gradient"
```
better to use this as an example of explicitly specify l2?

mathpluscode replied:
Sure

YipengHu reviewed (changes requested):
1 - for those removed custom files, please add tickets to add them back with updated register methods.
2 - add error message when recognising the old-format config files - i think it is pretty easy and safe (see my detailed comments).
3 - potential conflict with #561 and #54 should be easy to deal with, i am happy to merge this first then sync with @NMontanaBrown when she comes back.
4 - potential conflict with #525 and #559 should also be clear (and might be easier to write these docs after this PR), but please offer help when it needs to be merged to the @NMontanaBrown existing effort.

I do not want to keep blocking this, if this is going in, what @NMontanaBrown is working on needs to be merged anyway. If others are not available, consider ask @s-sd @zacbaum @kate-sann5100 to provide a second thorough review.

NMontanaBrown commented:
Hi @mathpluscode thanks for this, it is huge work and will add flexibility. I have returned from leave now, and will finish docs... Although I am not sure if they need refactoring/pulling into this branch before merge into main - I would suggest I finish the ticket as it is on my branch, then perhaps we can pull it into this one and review for new docs?

Although there is also a big question about back compatibility as has been raised already. Do we want duplicate docs? @zacbaum @YipengHu

mathpluscode commented on config/unpaired_labeled_ddf.yaml:
```diff
@@ -18,21 +18,16 @@ train:
     regularization:
       weight: 0.5 # weight of regularization loss
-      energy_type: "bending" # options include "bending", "gradient-l1" and "gradient-l2"
+      name: "bending" # options include "bending", "gradient"
```
I can put the default value there

mathpluscode replied:
oh this is bending, so no args.

mathpluscode commented:
> Hi @mathpluscode thanks for this, it is huge work and will add flexibility. I have returned from leave now, and will finish docs... Although I am not sure if they need refactoring/pulling into this branch before merge into main - I would suggest I finish the ticket as it is on my branch, then perhaps we can pull it into this one and review for new docs?
> 
> 
> 
> Although there is also a big question about back compatibility as has been raised already. Do we want duplicate docs? @zacbaum @YipengHu 

I would prefer merging my PR. And I can go to your branch modify the config docs :) what do you think @NMontanaBrown

mathpluscode commented:
> 1 - for those removed custom files, please add tickets to add them back with updated register methods.
> 
> 2 - add error message when recognising the old-format config files - i think it is pretty easy and safe (see my detailed comments).
> 
> 3 - potential conflict with #561 and #54 should be easy to deal with, i am happy to merge this first then sync with @NMontanaBrown when she comes back.
> 
> 4 - potential conflict with #525 and #559 should also be clear (and might be easier to write these docs after this PR), but please offer help when it needs to be merged to the @NMontanaBrown existing effort.
> 
> 
> 
> I do not want to keep blocking this, if this is going in, what @NMontanaBrown is working on needs to be merged anyway. If others are not available, consider ask @s-sd @zacbaum @kate-sann5100 to provide a second thorough review.

Yes will add tickets for the custom docs.

I do not like back compatibility as it makes the whole code super ugly. The key is refactoring config. So I will do more error messages.

For potential conflicts, as those branches are not merged. It's inevitable so I would contact the owners to offer help in case of necessity:)

🤔ok for you? @YipengHu

YipengHu commented:
> Hi @mathpluscode thanks for this, it is huge work and will add flexibility. I have returned from leave now, and will finish docs... Although I am not sure if they need refactoring/pulling into this branch before merge into main - I would suggest I finish the ticket as it is on my branch, then perhaps we can pull it into this one and review for new docs?
> 
> Although there is also a big question about back compatibility as has been raised already. Do we want duplicate docs? @zacbaum @YipengHu

sorry i wasn't clear. Yup - you can keep working/finishing what you have. When they are ready, we will see if it is easier to merge first or change per previously merged, e.g. this PR.

mathpluscode commented:
@YipengHu I've
- fixed some names so that 3D means we want the image to be 3D, without 3D means it can be anything. I added docs to those loss classes.
- added docstrings for mixin classes
- added a more meaningful error message in case of the wrong config.

Here's one problem that, as far as I know, it is hard to have back compatibility is that, now building class objects are centralized so that we do not know which class we are doing. It is different from changing the args of a certain class, where it is easier to have back compatibility.

YipengHu commented:
> @YipengHu I've
> - fixed some names so that 3D means we want the image to be 3D, without 3D means it can be anything. I added docs to those loss classes.
> - added docstrings for mixin classes
> - added a more meaningful error message in case of the wrong config.
> 
> Here's one problem that, as far as I know, it is hard to have back compatibility is that, now building class objects are centralized so that we do not know which class we are doing. It is different from changing the args of a certain class, where it is easier to have back compatibility.
> 
> Requesting back compatibility will lead to lot of if/else inside registry.... which is exactly against the design rule of the registry lol

Get rid of "3d" then. Only mention otherwise eg 2d.

mathpluscode commented:
> > @YipengHu I've
> > 
> > * fixed some names so that 3D means we want the image to be 3D, without 3D means it can be anything. I added docs to those loss classes.
> > * added docstrings for mixin classes
> > * added a more meaningful error message in case of the wrong config.
> > 
> > Here's one problem that, as far as I know, it is hard to have back compatibility is that, now building class objects are centralized so that we do not know which class we are doing. It is different from changing the args of a certain class, where it is easier to have back compatibility.
> > Requesting back compatibility will lead to lot of if/else inside registry.... which is exactly against the design rule of the registry lol
> 
> Get rid of "3d" then. Only mention otherwise eg 2d.

fine

mathpluscode commented:
@NMontanaBrown could you review this PR as we discussed ^^ I've double-checked the doc string of those losses.

YipengHu commented on deepreg/registry.py:
```diff
@@ -96,7 +96,15 @@ def build_from_config(self, category: str, config: dict, default_args=None):
+            return cls(**args)
+        except TypeError as err:
+            raise ValueError(
+                f"Configuration  is not compatible for Class {cls} of category {category}.\n"
```
Last comment:
i think it won't hurt to explicitly mention that "it is possible that an old format of configuration file may be being used."

and why two spaces between "Configuration" and "is"?

mathpluscode replied:
done

NMontanaBrown commented:
Reviewing now.

NMontanaBrown commented on CHANGELOG.md:
```diff
@@ -10,7 +10,7 @@ adheres to [Semantic Versioning](http://semver.org/).
 ### Added
 
-- Added registry for backbone models.
+- Added registry for backbone models and losses.
```
Since this is likely to raise a backcompatibility issue, I would add a link in the changelog to a documentation release.

mathpluscode replied:
what do you mean? we do not have docs...

NMontanaBrown replied:
We do not, but this is a major change so we should have docs. Could you write something quick for registry before merge?

mathpluscode replied:
registry stuff is not all done yet, i will do a doc when all is done. this is main not a release. otherwise you will be looking at a PR with 200 file changes...

mathpluscode replied:
so for docs, i will leave it to another PR/issue

NMontanaBrown commented on deepreg/dataset/loader/h5_loader.py:
```diff
@@ -143,4 +143,10 @@ def get_num_images(self) -> int:
+                try:
+                    f.close()
+                except RuntimeError:
+                    # to ignore RuntimeError: Can't increment id ref count
```
Is this the best handling case for a runtime error?

mathpluscode replied:
i got strange error that only occurs in CI not locally, I have no other methods :(

zacbaum replied:
Can we not at least log a warning saying that file did not close instead? Catching an error and not notifying the user isn't ideal.

NMontanaBrown replied:
Yes, I am a bit worried about the way we handle this, it could come back to bite back...? Better to investigate the root cause and solve (in a different ticket, perhaps) what is happening here.

mathpluscode replied:
ok i removed this change as a closer look tells me that i dont need to close files

NMontanaBrown commented on deepreg/model/loss/deform.py:
```diff
@@ -54,69 +56,82 @@ def gradient_dxyz(fxyz: tf.Tensor, fn: Callable) -> tf.Tensor:
+            norms = dfdx ** 2 + dfdy ** 2 + dfdz ** 2
+        return tf.reduce_mean(norms)
+
+    def get_config(self):
```
Docstring

mathpluscode replied:
added

mathpluscode replied:
added

NMontanaBrown commented on deepreg/model/loss/deform.py:
```diff
@@ -54,69 +56,82 @@ def gradient_dxyz(fxyz: tf.Tensor, fn: Callable) -> tf.Tensor:
-        raise ValueError(f"Unknown energy_type {energy_type}.")
+    def __init__(self, name="BendingEnergy"):
+        """
+        :param name: name of the loss
```
type missing

mathpluscode replied:
added, more of these things will be resolved with the pylint PR, as we need a tool to check them all.

NMontanaBrown commented on deepreg/model/loss/image.py:
```diff
@@ -3,214 +3,317 @@
+        name="GlobalMutualInformation",
+    ):
+        """
+        :param num_bins: number of bins for intensity
```
Can we add why num bins is default 23?

YipengHu replied:
This is probably beyond the topic of this PR, but it is an empirical default and reasonable to have a "balanced" number of bins - sufficiently distinct intensity values say for a 8bit image - and number of samples in each bin.

NMontanaBrown replied:
I would still add this explanation to the docstring. Even if it is empirically determined, magic numbers are annoying to have and manage in code.

mathpluscode replied:
I do not know... we can do a new issue to ask @acasamitjana

NMontanaBrown replied:
Can you add this to the docstring as I have suggested? For the magic number managing reason, and raise a ticket besides this to discuss if this needs changning.

mathpluscode replied:
done, https://github.com/DeepRegNet/DeepReg/issues/591

NMontanaBrown commented on deepreg/model/loss/image.py:
```diff
@@ -3,214 +3,317 @@
+        div = (pab + EPS) / (papb + EPS)
+        return tf.reduce_sum(pab * tf.math.log(div + EPS), axis=[1, 2])
+
+    def get_config(self):
```
docstring

mathpluscode replied:
added

NMontanaBrown commented on deepreg/model/loss/image.py:
```diff
@@ -3,214 +3,317 @@
+        name="LocalNormalizedCrossCorrelation",
+    ):
+        """
+        :param kernel_size: int. Kernel size or kernel sigma for kernel_type='gauss'.
```
Explanation for default kernel size?

mathpluscode replied:
size of kernel in convolution?

mathpluscode replied:
updated

NMontanaBrown commented on deepreg/model/loss/image.py:
```diff
@@ -3,214 +3,317 @@
+        ncc = (cross * cross + EPS) / (t_var * p_var + EPS)
+        return tf.reduce_mean(ncc, axis=[1, 2, 3, 4])
+
+    def get_config(self):
```
docstring

mathpluscode replied:
added

NMontanaBrown commented on deepreg/model/loss/label.py:
```diff
@@ -1,264 +1,313 @@
-    return (numerator + EPS) / (denominator + EPS)
+        return (y_prod + EPS) / (y_sum - y_prod + EPS)
 
+    def get_config(self):
```
docstring

mathpluscode replied:
added

NMontanaBrown commented on docs/source/docs/cli.md:
```diff
@@ -219,7 +219,7 @@ configuration can be specified in the configuration file. Please see
 
   - `--save_png`, for saving the outputs in png format.
-  - `--save_png`, for not saving the outputs in png format.
+  - `--no_png`, for not saving the outputs in png format.
```
why has this been changed? what does this have to do with losses?

mathpluscode replied:
it's a typo and i correct it together ^^

NMontanaBrown replied:
You need to add this to the changelog even if it's minor, to keep track, as it doesnt have it's own ticket. It's confusing that unrelated things get changed. We have discussed previously to separate issues into different tickets.

mathpluscode replied:
done

NMontanaBrown commented on docs/source/index.rst:
```diff
@@ -15,7 +15,7 @@
     tutorial/registration
     tutorial/cross_val
-    tutorial/custom
+    tutorial/run_cluster
```
What does this have to do with modifying losses + using registry? why has this been changed in this PR?

mathpluscode replied:
just correct small things together, they are minor stuff

NMontanaBrown replied:
Minor or not, same reason as above, we lose things in the fold - please add to changelog and in future raise separate ticket?

mathpluscode replied:
done

NMontanaBrown reviewed (changes requested):
I have left comments. Generally nice work, just some missing docstrings and also questions about some changes that seem unrelated to registry and losses in this PR...?

mathpluscode commented:
> I have left comments. Generally nice work, just some missing docstrings and also questions about some changes that seem unrelated to registry and losses in this PR...?

I've added missing docstrings for `get_config` and improve some other docs.

Regarding those small changes, it's just to fix some side issues altogether, I consider them as minor ^^.
One is related to typo and one is related to link a page, so I won't add typo in the changelog and add docs page is already included in "Simplified contributing documentation".

mathpluscode commented:
ok i solved the comments, @NMontanaBrown, for the docs: as mentioned (maybe multiple times) I will do a doc in the end when I'm 100% sure about the registry, I still have multiple PRs to do, and this is also a reason why we have `main` branch, which is considered as unstable by default ^^

NMontanaBrown commented:
You mention registry is not done, what is missing for registry? If we are not sure about it as a feature 100%, we should not be merging it into the main branch and it should be maintained as a WIP PR.

Irregardless I still think it is a very big change from our current working pipeline so explanations about it should be added in some form of documentation WITH the merge, whenever it happens, not separately, as this is a non-intuitive, non-minor change.

mathpluscode commented:
> You mention registry is not done, what is missing for registry? If we are not sure about it as a feature 100%, we should not be merging it into the main branch and it should be maintained as a WIP PR.
> 
> Irregardless I still think it is a very big change from our current working pipeline so explanations about it should be added in some form of documentation WITH the merge, whenever it happens, not separately, as this is a non-intuitive, non-minor change.

The registry for now is not for users, and users do not need to know anything with registry, they only need to provide correct configs. So i dont know what kind of docs you would like to have. The docs we need will be in your branch.

When I say ready, i mean user will be able to use the registry with their custom classes. For this we have other issues open already.

I need to merge this as it is now blocking many things ^^ and tbh it's not a dirty merge.

mathpluscode commented:
@NMontanaBrown if you can clarify a bit more on the docs you would like to have?

YipengHu commented:
> @NMontanaBrown if you can clarify a bit more on the docs you would like to have?

@NMontanaBrown @mathpluscode 

First, the changelog has been updated. From the PR, what I can see needs documentation support is mainly due to the format change of the config files. This was why I brought up the potential conflict with the Nina's ticket. While there is no reason doc and code need to be in a single PR, if the existing doc is non-existent or insufficient - there are plenty of undocumented features in every software. One solution would be merge #525 first to this PR (this might be perceived to reduce the chance of further tech debt, but also has a price - more on this later). If #525 can be done quickly, then this is definitely an option nonetheless. @mathpluscode what is specifically being blocked by this PR and cannot merge this branch to first instead? 

For compatibility issue for the config change, several of us agreed that the added error messages is the best to mitigate the risk of confusing users. We discussed this in length, trying to balance between a very wordy tedious explanation to "old users" and simplicity-and-clarity. Personally, I believe it is adequate for now. And the fact that the changed config is more structured, simplified and clearer now, with #525, we should have a better overall presented config. @NMontanaBrown Do you have any other specific suggestion to further improve it?

Even the answers to both above questions are nos, I suggest to merge this asap, simplify due to "the principle of being agile and snappy" ;) As we discussed many times, another type of tech debt will build up quickly too, for example every newly created branch will need to rebase this big PR before merge.

NMontanaBrown commented:
> @NMontanaBrown if you can clarify a bit more on the docs you would like to have?

Ideally we need proper API docs and a page in the docs for registry - this will take more time and is the complete solution and we need to do it so we should raise a ticket before package release or do it now. Ideally, do it now, because that way we provide a complete, and not half baked code solution into the main branch without documentation on how to use new functionality.

For this PR, there should **at least** be a .md file that at least explains nominally WHY we have changed from classes -> registry workflow, and also provide some pointers on how this works so nominally users will be able to debug their own code if they are coding from main branch and using config files that will no longer be supported.

mathpluscode commented:
> > @NMontanaBrown if you can clarify a bit more on the docs you would like to have?
> 
> Ideally we need proper API docs and a page in the docs for registry - this will take more time and is the complete solution and we need to do it so we should raise a ticket before package release or do it now. Ideally, do it now, because that way we provide a complete, and not half baked code solution into the main branch without documentation on how to use new functionality. **@mathpluscode EDIT: I see this is a ticket at #585. Then a release with a small .md on registry if fine**
> 
> For this PR, there should **at least** be a .md file that at least explains nominally WHY we have changed from classes -> registry workflow, and also provide some pointers on how this works so nominally users will be able to debug their own code if they are coding from main branch and using config files that will no longer be supported.

I've added the doc ^^ have a look

YipengHu commented on docs/source/docs/registry.md:
```diff
@@ -0,0 +1,31 @@
+dictionary mapping `(category, key)` to `value`. It also provides the
+`build_from_config` functionality, which allows creating a class instance from a config
+directly. This allows the simplification of the config so that each configuration file
+need to provide the name and necessary arguments.
```
needs

NMontanaBrown replied:
Can change line 18 to "This allows for the simplification of the config file such that each configuration only needs to provide the name of the class and the arguments to build the class for use. Documentation on Registry config will be released with issue #525 ."

Bit more comprehensive?

mathpluscode replied:
i added, except the last bit about doc, as it's mentioned in the title.

YipengHu commented on docs/source/docs/registry.md:
```diff
@@ -0,0 +1,31 @@
+> This is a functionality under active development. A full documentation will be
+> released later.
+
+DeepReg is adapting the usage of the registry system recently, to facilitate the usage
```
adopting?

YipengHu commented on docs/source/docs/registry.md:
```diff
@@ -0,0 +1,31 @@
+need to provide the name and necessary arguments.
+
+With the registry, when developing new classes **inside DeepReg**, we should use the
+corresponding decorator, so that the class is registered, cf.
```
why cf.?

NMontanaBrown replied:
I don't understand this either

mathpluscode replied:
https://en.wikipedia.org/wiki/Cf.#:~:text=The%20abbreviation%20cf.,with%20the%20topic%20being%20discussed.

YipengHu commented:
@mathpluscode @NMontanaBrown I see your registry page now - only minor typos, otherwise fine to me.

NMontanaBrown reviewed (commented):
Looks good, thanks @mathpluscode , agree with Yipeng, typo fix + 1 modif and approve.

mathpluscode commented:
@YipengHu @NMontanaBrown done ;) have another look plz
